### PR TITLE
feat(ingestion): Glue stateful ingestion

### PR DESCRIPTION
**Pull request by amanda-her**
## Checklist
- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/aws/glue.py [hunk 1]
```diff
@@ -1189,5 +1291,56 @@ def get_data_platform_instance() -> DataPlatformInstanceClass:
...
+    def create_checkpoint(self, job_id: JobId) -> Optional[Checkpoint]:
+        """
+        Create the custom checkpoint with empty state for the job.
+        """
+        assert self.ctx.pipeline_name is not None
+        if job_id == self.get_default_ingestion_job_id():
+            return Checkpoint(
+                job_name=job_id,
+                pipeline_name=self.ctx.pipeline_name,
+                platform_instance_id=self.get_platform_instance_id(),
```
Comment (shirshanka) [inline metadata-ingestion/src/datahub/ingestion/source/aws/glue.py:1303]: It is not clear to me why a platform instance is mandatory for stateful ingestion for glue

File: metadata-ingestion/src/datahub/ingestion/source/aws/glue.py [hunk 2]
```diff
@@ -1189,5 +1291,56 @@ def get_data_platform_instance() -> DataPlatformInstanceClass:
...
+                run_id=self.ctx.run_id,
+                config=self.source_config,
+                # Reusing BaseSQLAlchemyCheckpointState as it has needed functionality to support statefulness of GLue
+                state=BaseSQLAlchemyCheckpointState(),
+            )
+        return None
+
+    def get_platform_instance_id(self) -> str:
+        assert self.source_config.platform_instance is not None
```
Comment (shirshanka) [inline metadata-ingestion/src/datahub/ingestion/source/aws/glue.py:1312]: same comment: why is this mandatory?

File: metadata-ingestion/src/datahub/ingestion/source/aws/glue.py [hunk 3]
```diff
@@ -164,23 +201,28 @@ def platform_validator(cls, v: str) -> str:
...
+        self.soft_deleted_stale_entities.append(urn)
+
 
 @platform_name("Glue")
 @config_class(GlueSourceConfig)
 @support_status(SupportStatus.CERTIFIED)
 @capability(SourceCapability.PLATFORM_INSTANCE, "Enabled by default")
 @capability(SourceCapability.DOMAINS, "Supported via the `domain` config field")
-class GlueSource(Source):
+@capability(SourceCapability.DELETION_DETECTION, "Enabled by default when stateful ingestion is turned on. Stateful Ingestion is available only when a Platform Instance is assigned to this source.")
```
Comment (shirshanka) [inline metadata-ingestion/src/datahub/ingestion/source/aws/glue.py:224]: unclear why platform instance is required for supporting stateful ingestion

**Review by shirshanka — CHANGES_REQUESTED** [review]
Overall looks great, but I think you should be able to make platform_instance an optional requirement not mandatory?

File: metadata-ingestion/src/datahub/ingestion/source/aws/glue.py (hunk 2 above)
Comment (amanda-her) [reply metadata-ingestion/src/datahub/ingestion/source/aws/glue.py]: The idea of this method is to obtain the source identifier. If there are two sources with the same platform, the property `platform_instance` is what make them identifiable. However, to make stateful ingestion more flexible I have changed the implementation of this method making platform instance optional.

**Review by shirshanka — APPROVED** [review]
LGTM

**Pull request merged**